Source/GSOrderedSet.m

```diff
@@ -0,0 +1,358 @@
+    }
+  return self;
+}
+
+- (id) nextObject
+{
+  GSIArrayItem item;
+
+  if (current == 0)
+    return nil;
+  
+  item = GSIArrayItemAtIndex(&set->array, --current);
+  return (id)(item.obj);
+}
+@end
+
+@implementation GSOrderedSet
+
+static Class	setClass;
+static Class	mutableSetClass;
+
++ (void) initialize
+{
+  if (self == [GSOrderedSet class])
+    {
+      setClass = [GSOrderedSet class];
+      mutableSetClass = [GSMutableOrderedSet class];
+    }
+}
+
+- (id) copyWithZone: (NSZone*)z
+{
+  return RETAIN(self);
+}
+
+- (void) dealloc
+{
+  GSIArrayEmpty(&array);
+  [super dealloc];
+}
+
+- (NSUInteger) hash
+{
+  return [self count];
+}
+
+- (instancetype) init
+{
+  return [self initWithObjects: NULL count: 0];
+}
+
+- (NSEnumerator*) objectEnumerator
+{
+  return AUTORELEASE([[GSOrderedSetEnumerator alloc] initWithOrderedSet: self]);
+}
+
+- (NSEnumerator*) reverseObjectEnumerator
+{
+  return AUTORELEASE([[GSOrderedSetEnumeratorReverse alloc] initWithOrderedSet: self]);
+}
+
+- (NSUInteger) sizeInBytesExcluding: (NSHashTable*)exclude
+{
+  NSUInteger	size = GSPrivateMemorySize(self, exclude);
+
+  if (size > 0)
+    {
+      NSUInteger count = [self count];
+      NSUInteger i = 0;
+
+      for(i = 0; i < count; i++)
+	{
+	  GSIArrayItem item = GSIArrayItemAtIndex(&array, i);
+          size += [item.obj sizeInBytesExcluding: exclude];
+        }
+    }
+  return size;
+}
+
+// Put required overrides here...
+- (NSUInteger) count
+{
+  return GSIArrayCount(&array);
+}
+
+- (id) objectAtIndex: (NSUInteger)index
+{
+  GSIArrayItem item = GSIArrayItemAtIndex(&array, index);
+  return item.obj;
+}
+
+/* Designated initialiser */
+- (id) initWithObjects: (const id*)objs count: (NSUInteger)c
+{
+  NSUInteger i = 0;
+
+  // Initialize and fill the set.
+  GSIArrayInitWithZoneAndCapacity(&array, [self zone], c);
+  for (i = 0; i < c; i++)
+    {
+      id obj = objs[i];
+      GSIArrayItem item;
+      
+      if (objs[i] == nil)
+	{
+	  DESTROY(self);
+	  [NSException raise: NSInvalidArgumentException
+		      format: @"Tried to init set with nil value"];
+	}
+
+      item.obj = obj;
+      if(![self containsObject: obj])
+	{
+	  GSIArrayAddItem(&array, item);
+	  RETAIN(obj);
+	}
+    }
+  return self;
+}
+
+@end
+
+@implementation GSMutableOrderedSet
+
++ (void) initialize
+{
+  if (self == [GSMutableOrderedSet class])
+    {
+      GSObjCAddClassBehavior(self, [GSOrderedSet class]);
+    }
+}
+
+- (void) addObject: (id)anObject
+{
+  GSIArrayItem item;
+
+  if (anObject == nil)
+    {
+      [NSException raise: NSInvalidArgumentException
+		  format: @"Tried to add nil to set"];
+    }
+
+  if([self containsObject: anObject] == NO)
+    {
+      item.obj = anObject;
+      GSIArrayAddItem(&array, item);
+      RETAIN(anObject);
+      _version++;
+    }
+}
+
+- (void) insertObject: (id)object atIndex: (NSUInteger)index
+{
+  GSIArrayItem item;
+  if(object == nil)
+    {
+      NSWarnLog(@"Attempt to insert nil object");
```

Comment:
Why does this give a warning log, while addObject: raises an exception?

Reply:
This has been changed.